### ISSUE TITLE
RPCs not routed to different deepstream node?

### ISSUE THREAD

**Issue by TimMun** [deepstreamIO/deepstream.io#1001]
Not sure if this is expected behavior. I have the following setup: DataProvider A provides RPCs (using client.rpc.provide) and is connected to DeepstreamServer A. DataProvider B is connected to DeepstreamServer B. DeepstreamServer A and DeepstreamServer B are part of the same cluster (using deepstream.io-clusternode-redis). 

I have a need for DataProvider B to make an rpc to DataProvider A. However, when I do this, DataProvider B gets a "NO_RPC_PROVIDER" error. Is this expected? Can RPCs get routed across data providers connected to different server nodes?

(When I have both DataProvider A and B connected to DeepstreamServer A, then DataProvider B can successfully make the rpc to DataProvider A).

**Update**: According to [this](https://deepstream.io/tutorials/core/request-response/), rpcs should be routed across the cluster. After a little experimentation, the rpc routing only works when **both** DeepstreamServer A and B are online when DataProvider A starts providing.

So to summarize:

If the startup order goes:
1. DeepstreamServer A starts
2. DataProvider A login to DeepstreamServer A, starts providing rpc
3. DeepstreamServer B starts
4. DataProvider B login to DeepstreamServer B, then DataProvider B makes an rpc request to DataProvider A, it gets NO_RPC_PROVIDER error.

In contrast, if the startup order is: 
1. DeepstreamServer A starts
2. DeepstreamServer B starts
3. DataProvider A login to DeepstreamServer A, starts providing rpc
4. DataProvider B login to DeepstreamServer B, then DataProvider B makes an rpc request to DataProvider A; rpc will be succcessful.

**Comment by yasserf** [comment]
Interesting! Will take a look

**Comment by TimMun** [comment]
From my limited understanding of how DS works internally, I think the problem is that `onServerAdded` in DistributedStateRegistry is not ever being called. This causes the second DS node to never request the full state of the rpc registry.

**Comment by yasserf** [comment]
hey, looked into this and you might be right, although that code was there from some older logic that didn't get cleaned up. I need to do some more digging into this. As chance you have a repo to reproduce on you? Got quite a list of things to do so every little helps. 

Thanks for raising btw!

**Comment by TimMun** [comment]
Sure, here it is:

https://github.com/TimMun/deepstream-rpc-bug

To start, run "node failure.js".

**Comment by yasserf** [comment]
Thank you!

**Comment by yasserf** [comment]
Sorry for not getting on this yet, it’s one of my top properties, things are just a little hectic right now. Will aim to get this out in a new release in the next couple of days as I’ll be house bound 😅

**Comment by yasserf** [comment]
Also events work due to the broadcast mechanism in redis cluster just blasting it out to all servers regardless. Be interesting to see if listening works, I will be adding this as an e2e test including what happens with different permutations to see what happens

**Comment by yasserf** [comment]
Fixed in v4.2.3

Thanks again for raising!

Test here: https://github.com/deepstreamIO/deepstream.io-e2e/blob/master/rpcs/rpc-reconnectivity.feature#L51
